File: app/src/androidTest/java/name/boyle/chris/sgtpuzzles/ChooserScreenshotsTest.java
package name.boyle.chris.sgtpuzzles;

import static android.content.Context.MODE_PRIVATE;
import static androidx.test.core.app.ApplicationProvider.getApplicationContext;
import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.scrollTo;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isCompletelyDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withChild;
import static androidx.test.espresso.matcher.ViewMatchers.withEffectiveVisibility;
import static androidx.test.espresso.matcher.ViewMatchers.withSubstring;
import static name.boyle.chris.sgtpuzzles.PrefsConstants.CHOOSER_STYLE_KEY;
import static name.boyle.chris.sgtpuzzles.PrefsConstants.STATE_PREFS_NAME;

import android.content.SharedPreferences;
import android.os.SystemClock;

import androidx.preference.PreferenceManager;
import androidx.test.espresso.matcher.ViewMatchers;
import androidx.test.ext.junit.rules.ActivityScenarioRule;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import tools.fastlane.screengrab.Screengrab;
import tools.fastlane.screengrab.cleanstatusbar.CleanStatusBar;

@RunWith(JUnit4.class)
public class ChooserScreenshotsTest {

    private final static SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
    private final static SharedPreferences state = getApplicationContext().getSharedPreferences(STATE_PREFS_NAME, MODE_PRIVATE);

    @BeforeClass
    public static void beforeAll() {
        CleanStatusBar.enableWithDefaults();
        prefs.edit().clear().apply();
        state.edit().clear().apply();
    }

    @AfterClass
    public static void afterAll() {
        CleanStatusBar.disable();
    }

    @Rule
    public final ActivityScenarioRule<GameChooser> activityRule = new ActivityScenarioRule<>(GameChooser.class);

    @Test
    public void testTakeScreenshots() {
        switch (TestUtils.getFastlaneDeviceTypeOrSkipTest()) {
            case "tenInch" -> {
                scrollToAndScreenshot(GUESS.INSTANCE, "07_chooser");
                scrollToAndScreenshot(UNTANGLE.INSTANCE, "08_chooser");
            }
            case "sevenInch" -> {
                scrollToAndScreenshot(GUESS.INSTANCE, "06_chooser");
                scrollToAndScreenshot(SAMEGAME.INSTANCE, "07_chooser");
                scrollToAndScreenshot(UNTANGLE.INSTANCE, "08_chooser");
            }
            case "phone" -> {
                prefs.edit().putString(CHOOSER_STYLE_KEY, "grid")
                        // Star 3 arbitrarily chosen extra games in order that all 40 games fit on a 16x9 screen
                        .putBoolean("starred_" + LOOPY.INSTANCE, true)
                        .putBoolean("starred_" + MOSAIC.INSTANCE, true)
                        .putBoolean("starred_" + TRACKS.INSTANCE, true)
                        .apply();
                onView(withSubstring(GUESS.INSTANCE.getDisplayName())).check(matches(withEffectiveVisibility(ViewMatchers.Visibility.GONE)));
                SystemClock.sleep(100);  // Espresso thinks we're idle before the animation has finished :-(
                assertIconVisible(GUESS.INSTANCE);
                assertIconVisible(UNTANGLE.INSTANCE);
                Screengrab.screenshot("08_chooser_grid");
                prefs.edit().clear().apply();
            }
        }
    }

    private void assertIconVisible(BackendName backend) {
        onView(withChild(withSubstring(backend.getDisplayName())))
                .check(matches(isCompletelyDisplayed()));
    }

    private void scrollToAndScreenshot(final BackendName backend, final String screenshotName) {
        onView(withChild(withSubstring(backend.getDisplayName())))
                .perform(scrollTo())
                .check(matches(isCompletelyDisplayed()));
        SystemClock.sleep(100);  // Espresso thinks we're idle before the scroll has finished :-(
        Screengrab.screenshot(screenshotName);
    }
}

File: app/src/androidTest/kotlin/name/boyle/chris/sgtpuzzles/ChooserScreenshotsTest.kt
package name.boyle.chris.sgtpuzzles

import android.content.Context
import android.os.SystemClock
import androidx.preference.PreferenceManager
import androidx.test.core.app.ApplicationProvider
import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.ext.junit.rules.ActivityScenarioRule
import name.boyle.chris.sgtpuzzles.backend.BackendName
import name.boyle.chris.sgtpuzzles.backend.GUESS
import name.boyle.chris.sgtpuzzles.backend.LOOPY
import name.boyle.chris.sgtpuzzles.backend.MOSAIC
import name.boyle.chris.sgtpuzzles.backend.SAMEGAME
import name.boyle.chris.sgtpuzzles.backend.TRACKS
import name.boyle.chris.sgtpuzzles.backend.UNTANGLE
import name.boyle.chris.sgtpuzzles.config.PrefsConstants.CHOOSER_STYLE_KEY
import name.boyle.chris.sgtpuzzles.config.PrefsConstants.STATE_PREFS_NAME
import org.junit.AfterClass
import org.junit.BeforeClass
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import tools.fastlane.screengrab.Screengrab
import tools.fastlane.screengrab.cleanstatusbar.CleanStatusBar

@RunWith(JUnit4::class)
class ChooserScreenshotsTest {
    @JvmField
    @Rule
    val activityRule = ActivityScenarioRule(
        GameChooser::class.java
    )

    @Test
    fun testTakeScreenshots() {
        when (TestUtils.fastlaneDeviceTypeOrSkipTest) {
            "tenInch" -> {
                scrollToAndScreenshot(GUESS, "07_chooser")
                scrollToAndScreenshot(UNTANGLE, "08_chooser")
            }

            "sevenInch" -> {
                scrollToAndScreenshot(GUESS, "06_chooser")
                scrollToAndScreenshot(SAMEGAME, "07_chooser")
                scrollToAndScreenshot(UNTANGLE, "08_chooser")
            }

            "phone" -> {
                prefs.edit().putString(CHOOSER_STYLE_KEY, "grid")
                    // Star 3 arbitrarily chosen extra games in order that all 40 games fit on a 16x9 screen
                    .putBoolean("starred_$LOOPY", true)
                    .putBoolean("starred_$MOSAIC", true)
                    .putBoolean("starred_$TRACKS", true)
                    .apply()
                Espresso.onView(ViewMatchers.withSubstring(GUESS.displayName))
                    .check(ViewAssertions.matches(ViewMatchers.withEffectiveVisibility(ViewMatchers.Visibility.GONE)))
                SystemClock.sleep(100) // Espresso thinks we're idle before the animation has finished :-(
                assertIconVisible(GUESS)
                assertIconVisible(UNTANGLE)
                Screengrab.screenshot("08_chooser_grid")
                prefs.edit().clear().apply()
            }
        }
    }

    private fun assertIconVisible(backend: BackendName) {
        Espresso.onView(ViewMatchers.withChild(ViewMatchers.withSubstring(backend.displayName)))
            .check(ViewAssertions.matches(ViewMatchers.isCompletelyDisplayed()))
    }

    private fun scrollToAndScreenshot(backend: BackendName, screenshotName: String) {
        Espresso.onView(ViewMatchers.withChild(ViewMatchers.withSubstring(backend.displayName)))
            .perform(ViewActions.scrollTo())
            .check(ViewAssertions.matches(ViewMatchers.isCompletelyDisplayed()))
        SystemClock.sleep(100) // Espresso thinks we're idle before the scroll has finished :-(
        Screengrab.screenshot(screenshotName)
    }

    companion object {
        private val prefs =
            PreferenceManager.getDefaultSharedPreferences(ApplicationProvider.getApplicationContext())
        private val state = ApplicationProvider.getApplicationContext<Context>()
            .getSharedPreferences(STATE_PREFS_NAME, Context.MODE_PRIVATE)

        @JvmStatic
        @BeforeClass
        fun beforeAll() {
            CleanStatusBar.enableWithDefaults()
            prefs.edit().clear().apply()
            state.edit().clear().apply()
        }

        @JvmStatic
        @AfterClass
        fun afterAll() {
            CleanStatusBar.disable()
        }
    }
}